service/controller/azurecluster/handler/azureclusteridentity/create.go

```diff
@@ -0,0 +1,198 @@
+package azureclusteridentity
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v5/pkg/label"
+)
+
+const (
+	secretDataFieldName = "clientSecret"
+)
+
+// EnsureCreated ensures there is an AzureClusterIdentity CR and a related Secret
+// with the same contents as the Giant Swarm credential secret.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	legacySecrets, err := r.listLegacySecrets(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, legacySecret := range legacySecrets {
+		r.logger.Debugf(ctx, "Found secret %q in namespace %q", legacySecret.Name, legacySecret.Namespace)
+
+		err = r.ensureNewSecret(ctx, legacySecret)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = r.ensureAzureClusterIdentity(ctx, legacySecret)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
```

Comment:
Instead of always processing all credentials, would it make sense to only process the ones that are in use? More specifically, only ensure `AzureClusterIdentity` for the credential that reconciled CR is actually using? This would reduce number of unnecessary API calls from the k8s apiserver. In present state this is listing all credentials and invoking two separate GETs for each legacy credential, performed on every CRs reconciliation loop.

This could be roughly something like following:
- Check if reconciled `AzureCluster` has `IdentityRef` filled. If yes, bail out early (no further actions).
- When `IdentityRef` is not populated, check if `AzureClusterIdentity` for given legacy credential exists.
- If not, proceed ensuring that related new secret & azure cluster identity CR exists.
- Update reconciled CR's `IdentityRef` with given `AzureClusteridentity`.

WDYT?

Reply:
That sounds like a good approach.

Reply:
Doing that would already solve https://github.com/giantswarm/giantswarm/issues/15576

Reply:
on it

Reply:
done